Add a VDE backend driver for virtio-net

tommie-lie opened the pull request:
This PR adds a new lkl_netdev device implementation that uses Virtualsquare's [VDE](http://vde.sourceforge.net/) userspace/virtual network.

Direct VDE support in LKL allows applications to directly connect to VDE networks without implementing their own TCP/IP stack or relying on the host system support (root permissions + tap device). This comes in handy if "hosts" in the virtual network are single-purpose applications (like a DHCP server, a SOCKS proxy, ...) that don't justify the overhead of a full-size virtual machine.

To build with the VDE network backend, you have to have libvdeplug2-dev installed and use `make vde=y`.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/lkl/linux/122)

<!-- Reviewable:end -->

thehajime commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[tools/lkl/bin/lkl-hijack.sh, line 18 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYRcNWsGONANID-_JJ:-KCYRcNXtpKj5PF9_tJL:1716752984) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/bin/lkl-hijack.sh#L18)):*
is it possible to avoid using hard-coded path here ?
it would be nice if we can take path from some standard way like pkg-config output.

---

*[tools/lkl/lib/virtio_net_vde.c, line 1 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYRpc4bz9j7YHKHblv:-KCYRpc4bz9j7YHKHblw:-1180001175) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/lib/virtio_net_vde.c#L1)):*
I think I will follow this way in my DPDK patch.

---

*[tools/lkl/lib/virtio_net_vde.c, line 39 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYRz2gzLTsIiQsfw3B:-KCYRz2h1IB7k37w0ZDs:-181799611) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/lib/virtio_net_vde.c#L39)):*
```EAGAIN``` instead of  the negative value ?
(I also had the same comment in my tap patch).

---

*[tools/lkl/tests/hijack-test.sh, line 29 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYSA2IBtAieYhStay-:-KCYSA2IBtAieYhStay0:329798443) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/tests/hijack-test.sh#L29)):*
is this commented-out line needed for some specific reason of vde or just an error ?

---

*[tools/lkl/tests/hijack-test.sh, line 32 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYSI3rFTzeDbIif2-B:-KCYSI3rFTzeDbIif2-C:-1677837463) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/tests/hijack-test.sh#L32)):*
ditto.

---

*[tools/lkl/tests/hijack-test.sh, line 132 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYSMqOkkQNoAhFoneU:-KCYSMqOkkQNoAhFoneV:-1147186957) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/tests/hijack-test.sh#L132)):*
I guess those clean up lines can be move to trap handler (function clear_work_dir).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/122)*
<!-- Sent from Reviewable.io -->

pscollins commented:
This looks great!

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[tools/lkl/bin/lkl-hijack.sh, line 18 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYRcNWsGONANID-_JJ:-KCanrhag6toeOZVS6sQ:755843727) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/bin/lkl-hijack.sh#L18)):*
+1

Also I think this shouldn't be loaded unless LKL has been compiled with VDE support -- it's already tricky to debug issues that arise during the `LD_PRELOAD` phase.

Maybe drop a symlink to `libvdevplug.so` in `lib` if we've been compiled with VDE support, and allow a different libdevplug path to be specified during the build? Then `LD_PRELOAD` the symlink if it exists in `lkl-hijack.sh`

---

*[tools/lkl/lib/hijack/init.c, line 171 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCapBcYbm75_75icbYN:-KCapBcZsTvsCMORxFHY:366013453) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/lib/hijack/init.c#L171)):*
Linux style guide requires { } if there is an `else` statement, I believe

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/122#-:-KCapZNyFm7HnhJIBCzV:-1155845411)*
<!-- Sent from Reviewable.io -->

tommie-lie commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[tools/lkl/bin/lkl-hijack.sh, line 18 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYRcNWsGONANID-_JJ:-KCb8E7bz6khJ2QCzGma:-447406211) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/bin/lkl-hijack.sh#L18)):*
As liblkl-hijack.so has proper link flags now, this is unnecessary, anyway. I will just remove the additional `LD_PRELOAD`.

---

*[tools/lkl/lib/virtio_net_vde.c, line 39 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYRz2gzLTsIiQsfw3B:-KCb6Vi8DPaN8fw88Wwv:-264375008) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/lib/virtio_net_vde.c#L39)):*
Copy&paste error, will fix :-)

---

*[tools/lkl/lib/hijack/init.c, line 171 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCapBcYbm75_75icbYN:-KCb6cw43QWTgQ2Cz_qQ:-1559840695) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/lib/hijack/init.c#L171)):*
Not according to https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/CodingStyle#n156

---

*[tools/lkl/tests/hijack-test.sh, line 29 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYSA2IBtAieYhStay-:-KCb68xOHYOakWo63Cdd:-108806732) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/tests/hijack-test.sh#L29)):*
Damn, this is a remains of debugging, I was sure I un-commented them again.

---

*[tools/lkl/tests/hijack-test.sh, line 132 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCYSMqOkkQNoAhFoneU:-KCb9KQMpYsR_GgRf7as:-199647383) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/tests/hijack-test.sh#L132)):*
Did not look too thoroughly at the new script after merging. Mea culpa.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/122)*
<!-- Sent from Reviewable.io -->

pscollins commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[tools/lkl/lib/hijack/init.c, line 171 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/122#-KCapBcYbm75_75icbYN:-KCbAXJ2bRl62NB_ZJ0Z:1176852643) ([raw file](https://github.com/lkl/linux/blob/a7fbd0c8b0ff143ffff7da7f04a1320199f57964/tools/lkl/lib/hijack/init.c#L171)):*
I stand corrected --- I was thinking of the "This does not apply if only one branch of a conditional statement is a single
statement;" line.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/122)*
<!-- Sent from Reviewable.io -->

tommie-lie commented:
I addressed all the comments from reviewable.io and rebased to the latest master (which already includes DPDK) and renamed my variables to match those of DPDK.
Please take a special look at [lib/hijack/init.c:168 ff.](https://github.com/lkl/linux/pull/122/files#diff-f9502b3c66b86f12248863c8b11d0ddfR168). I'm pretty sure the previous use of a short `strncmp` was unintentional and as we're comparing literals, `strcmp` should be used, anyway.

thehajime commented:
Reviewed 7 of 7 files at r1, 1 of 3 files at r2.
Review status: 1 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/122)*
<!-- Sent from Reviewable.io -->

thehajime commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/122)*
<!-- Sent from Reviewable.io -->

thehajime commented:
LGTM.
